Tests/AsyncHTTPClientTests/AsyncTestHelpers.swift

```diff
@@ -0,0 +1,189 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if swift(>=5.5) && canImport(_Concurrency)
+import NIOConcurrencyHelpers
+import NIOCore
+
+@available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
+class AsyncSequenceWriter: AsyncSequence {
+    typealias AsyncIterator = Iterator
+    typealias Element = ByteBuffer
+
+    struct Iterator: AsyncIteratorProtocol {
+        typealias Element = ByteBuffer
+
+        private let writer: AsyncSequenceWriter
+
+        init(_ writer: AsyncSequenceWriter) {
+            self.writer = writer
+        }
+
+        mutating func next() async throws -> ByteBuffer? {
+            try await self.writer.next()
+        }
+    }
+
+    func makeAsyncIterator() -> Iterator {
+        return Iterator(self)
+    }
+
+    private enum State {
+        case buffering(CircularBuffer<ByteBuffer?>, CheckedContinuation<Void, Error>?)
```

Comment:
Do we need to know that this is a ByteBuffer? Otherwise I would prefer it to be generic to reduce the accessible API surface inside the implementation of `AsyncSequenceWriter`.